plugins/discovery-ec2/qa/amazon-ec2/src/test/java/org/elasticsearch/discovery/ec2/AmazonEC2Fixture.java

```diff
@@ -16,106 +16,166 @@
         }
-        MockSecureSettings secureSettings = new MockSecureSettings();
-        secureSettings.setString(AwsEc2Service.ACCESS_KEY_SETTING.getKey(), "some_access");
-        secureSettings.setString(AwsEc2Service.SECRET_KEY_SETTING.getKey(), "some_secret");
-        return Settings.builder().put(super.nodeSettings(nodeOrdinal))
-            .put(DiscoveryModule.DISCOVERY_HOSTS_PROVIDER_SETTING.getKey(), "ec2")
-            .put("path.logs", resolve)
-            .put("transport.tcp.port", 0)
-            .put("node.portsfile", "true")
-            .put(AwsEc2Service.ENDPOINT_SETTING.getKey(), "http://" + httpServer.getAddress().getHostName() + ":" +
-                httpServer.getAddress().getPort())
-            .setSecureSettings(secureSettings)
-            .build();
     }
 
-    /**
-     * Creates mock EC2 endpoint providing the list of started nodes to the DescribeInstances API call
-     */
-    @BeforeClass
-    public static void startHttpd() throws Exception {
-        logDir = createTempDir();
-        httpServer = MockHttpServer.createHttp(new InetSocketAddress(InetAddress.getLoopbackAddress().getHostAddress(), 0), 0);
-
-        httpServer.createContext("/", (s) -> {
-            Headers headers = s.getResponseHeaders();
-            headers.add("Content-Type", "text/xml; charset=UTF-8");
-            String action = null;
-            for (NameValuePair parse : URLEncodedUtils.parse(IOUtils.toString(s.getRequestBody()), StandardCharsets.UTF_8)) {
-                if ("Action".equals(parse.getName())) {
-                    action = parse.getValue();
-                    break;
+    @SuppressForbidden(reason = "Paths#get is fine - we don't have environment here")
+    private static Path toPath(final String dir) {
+        return Paths.get(dir);
+    }
+
+    private static void writeFile(final Path dir, final String fileName, final String content) throws IOException {
+        final Path tempPidFile = Files.createTempFile(dir, null, null);
+        Files.write(tempPidFile, singleton(content));
+        Files.move(tempPidFile, dir.resolve(fileName), StandardCopyOption.ATOMIC_MOVE);
+    }
+
+    private static String addressToString(final SocketAddress address) {
+        final InetSocketAddress inetSocketAddress = (InetSocketAddress) address;
+        if (inetSocketAddress.getAddress() instanceof Inet6Address) {
+            return "[" + inetSocketAddress.getHostString() + "]:" + inetSocketAddress.getPort();
+        } else {
+            return inetSocketAddress.getHostString() + ":" + inetSocketAddress.getPort();
+        }
+    }
+
+    static class ResponseHandler implements HttpHandler {
+
+        private final Path discoveryPath;
+
+        ResponseHandler(final Path discoveryPath) {
+            this.discoveryPath = discoveryPath;
+        }
+
+        @Override
+        public void handle(HttpExchange exchange) throws IOException {
+            RestStatus responseStatus = RestStatus.INTERNAL_SERVER_ERROR;
```

Comment:
should this be `BAD_REQUEST` instead? If we cannot handle the request, it's not that there's an internal server error, but that the request is unexpected.

Reply:
Yes, BAD_REQUEST is better. It has been implemented in #31210 on top of which this PR has been rebased.